src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -1,299 +1,359 @@
+    private val BlockHeadersHandlerName = "block-headers-request-handler"
 
-    private var mptNodesQueue: Seq[HashType] = initialSyncState.mptNodesQueue
-    private var nonMptNodesQueue: Seq[HashType] = initialSyncState.nonMptNodesQueue
-    private var blockBodiesQueue: Seq[ByteString] = initialSyncState.blockBodiesQueue
-    private var receiptsQueue: Seq[ByteString] = initialSyncState.receiptsQueue
-    private var downloadedNodesCount: Int = initialSyncState.downloadedNodesCount
-    private var bestBlockHeaderNumber: BigInt = initialSyncState.bestBlockHeaderNumber
+    private var syncState = initialSyncState
 
     private var assignedHandlers: Map[ActorRef, Peer] = Map.empty
     private var peerRequestsTime: Map[Peer, Instant] = Map.empty
 
-    private val syncStateStorageActor = context.actorOf(Props[FastSyncStateActor], "state-storage")
-
     private var requestedMptNodes: Map[ActorRef, Seq[HashType]] = Map.empty
     private var requestedNonMptNodes: Map[ActorRef, Seq[HashType]] = Map.empty
     private var requestedBlockBodies: Map[ActorRef, Seq[ByteString]] = Map.empty
     private var requestedReceipts: Map[ActorRef, Seq[ByteString]] = Map.empty
 
+    private val syncStateStorageActor = context.actorOf(Props[FastSyncStateActor], "state-storage")
     syncStateStorageActor ! fastSyncStateStorage
 
-    private var blockChainOnlyPeers = Seq.empty[Peer]
+    private var blockChainOnlyPeers: Seq[Peer] = Nil
 
     //Delay before starting to persist snapshot. It should be 0, as the presence of it marks that fast sync was started
     private val persistStateSnapshotDelay: FiniteDuration = 0.seconds
-
     private val syncStatePersistCancellable = scheduler.schedule(persistStateSnapshotDelay, persistStateSnapshotInterval, self, PersistSyncState)
+    private val printStatusCancellable = scheduler.schedule(printStatusInterval, printStatusInterval, self, PrintStatus)
     private val heartBeat = scheduler.schedule(syncRetryInterval, syncRetryInterval * 2, self, ProcessSyncing)
 
-    // scalastyle:off cyclomatic.complexity
-    def receive: Receive = handlePeerUpdates orElse handleFailingMptPeers orElse {
-      case EnqueueNodes(hashes) =>
-        hashes.foreach {
-          case h: EvmCodeHash => nonMptNodesQueue = h +: nonMptNodesQueue
-          case h: StorageRootHash => nonMptNodesQueue = h +: nonMptNodesQueue
-          case h: StateMptNodeHash => mptNodesQueue = h +: mptNodesQueue
-          case h: ContractStorageMptNodeHash => mptNodesQueue = h +: mptNodesQueue
+    def receive: Receive = handleCommonMessages orElse {
+      case ProcessSyncing => processSyncing()
+      case PrintStatus => printStatus()
+      case PersistSyncState => persistSyncState()
+
+      case ResponseReceived(peer, BlockHeaders(blockHeaders), timeTaken) =>
+        log.info("Received {} block headers in {} ms", blockHeaders.size, timeTaken)
+        removeRequestHandler(sender())
+        handleBlockHeaders(peer, blockHeaders)
```

Comment:
it is not that easy to do as I initially thought. However, I'll try to come up with something